### PR TITLE
Game of life implementation

### PR DESCRIPTION
**Pull request by cdiielsi**
MVP for Game of Life implementation with a graphic interface developed. When executed (run `make`) the program should show a 50x50 grid with 3 patterns evolving. Game can be paused and unpaused hitting `space`. When paused you should be able to add new living cells or remove current living cells with `left click`. The README has the details about what the game of life is and how to run the program.

The cell evolution goes by the following rules:

- Any live cell with fewer than two live neighbours dies, as if by underpopulation.
- Any live cell with two or three live neighbours lives on to the next generation.
- Any live cell with more than three live neighbours dies, as if by overpopulation.
- Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,46 @@
...
+
+## Dependencies
+- rust 1.85.0
+- macroquad 0.4.14
+
+## References
+- https://en.wikipedia.org/wiki/Conway%27s_Game_of_Life
+- https://en.wikipedia.org/wiki/Cellular_automaton
+- https://macroquad.rs/docs/
+- https://github.com/not-fl3/macroquad/tree/master
```
Comment (Arkenan) [inline README.md:48]: I noticed some quotes from wikipedia while reading, great practice to list the articles as references! This will become specially important when reading code from other open source projects.

File: src/main.rs [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+use macroquad::prelude::*;
+
+mod gol;
+
+const WIDTH: usize = 50;
+const HEIGHT: usize = 50;
+
+#[macroquad::main("BasicShapes")]
+async fn main() {
+    let speed = 0.3;
```
Comment (Arkenan) [inline src/main.rs]: If this can't be configured, let's have it as a constant, same as width and height.

File: src/main.rs [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+        gol.draw_gol_board();
+
+        if is_key_pressed(KeyCode::Space) {
+            game_of_life_go = !game_of_life_go;
+        }
+
+        if !game_of_life_go && is_mouse_button_pressed(MouseButton::Left) {
+            let position = mouse_position();
+            gol.draw_or_drop_new_cell(position);
```
Comment (Arkenan) [inline src/main.rs]: nit: `draw_or_drop_new_cell` is fine, but you can also call it `toggle_cell`

File: src/main.rs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+#[macroquad::main("BasicShapes")]
+async fn main() {
+    let speed = 0.3;
+    let mut last_update = get_time();
+    let mut gol = GameOfLife::new(WIDTH, HEIGHT);
+    insert_3_cell_line_patter_top_left_corner(&mut gol);
+    insert_square_patter_top_left_corner(&mut gol);
+    insert_glider_patter_middle(&mut gol);
+    let mut game_of_life_go = true;
```
Comment (Arkenan) [inline src/main.rs]: nit: `game_of_life_go` is ok, but something like `is_game_running` or `is_game_paused` would be a bit clearer. It reads better with the ifs. e.g `if !is_game_running && is_mouse_pressed`.

File: src/gol.rs [hunk 5]
```diff
@@ -0,0 +1,314 @@
+use macroquad::prelude::*;
+use std::{cmp::min, collections::HashSet};
+
+/// Implementation of game of life. The struct has two field for the boards dimentions and anotherone to register
```
Comment (Arkenan) [inline src/gol.rs]:
```suggestion
/// Implementation of game of life. The struct has two fields for the boards dimentions and another one to register
```

File: src/gol.rs [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+pub struct GameOfLife {
+    width: usize,
+    height: usize,
+    alive_cells: HashSet<(usize, usize)>,
+}
+
+impl GameOfLife {
+    /// Creates an instance of a Game of Life with an empty set of alive cells and the dimentions of the board.
+    pub fn new(width: usize, height: usize) -> Self {
+        let alive_cells = HashSet::new();
```
Comment (Arkenan) [inline src/gol.rs]: nit: no need to have a variable for this. You can have 

```rust
Self {
  width,
  height,
  HashSet::new()
}

File: src/gol.rs [hunk 7]
```diff
@@ -0,0 +1,314 @@
+use macroquad::prelude::*;
+use std::{cmp::min, collections::HashSet};
+
+/// Implementation of game of life. The struct has two field for the boards dimentions and anotherone to register
+/// the living cells. After each transition this set is updated with the resulting living cells.
+pub struct GameOfLife {
+    width: usize,
+    height: usize,
+    alive_cells: HashSet<(usize, usize)>,
```
Comment (Arkenan) [inline src/gol.rs]: HashSet is ok, but given that the size is fixed once the game is created, have you considered a matrix/slice of slices? That would be more efficient in access times and also easier to get/update.

You don't need to change it now, but it would be good to have it documented in the readme as a design decision.

File: src/gol.rs [hunk 8]
```diff
@@ -0,0 +1,314 @@
...
+    pub fn new(width: usize, height: usize) -> Self {
+        let alive_cells = HashSet::new();
+        Self {
+            width,
+            height,
+            alive_cells,
+        }
+    }
+    /// Adds a cell to the alive cells set.
+    /// TODO: handle coordinates outiside the boards range with error
```
Comment (Arkenan) [inline src/gol.rs]: This is not optional, silent failure is not an option.

File: src/gol.rs (hunk 7 above)
Comment (Arkenan) [reply src/gol.rs]: Also, if you are going the hash set route you probably would benefit of a Cell {x: usize, y: usize} type. It's easier to pass around and also clearer when doing cell.x or cell.y instead of .0, .1.

File: src/gol.rs [hunk 9]
```diff
@@ -0,0 +1,314 @@
...
+    /// Updates alive_cell field with the new living cells.
+    pub fn transition(&mut self) -> Option<()> {
+        let mut next_iteration_set = HashSet::new();
+        for x in 0..self.width {
+            for y in 0..self.height {
+                let current_cell = (x, y);
+                self.process_cell(
+                    &mut next_iteration_set,
+                    current_cell,
+                    self.alive_cells.contains(&(x, y)),
```
Comment (Arkenan) [inline src/gol.rs]: The function process_cell may be doing a bit too much. "process" gives you the impression that it modifies the cell, but it actually modifies something different. I think it would be a better idea to have a `self.next_cell_state(cell)` and returns true if that cell is alive the next iteration. Note that you can get the current state inside the function. Then you can update the next_iteration_set with the result.

Thinking about this differently, the core logic of GoL is if a cell is alive or not next iteration based on its neighbors. If there's an iteration set or stuff like that is a bit of an implementation detail and it's good to separate that from that core function.

File: src/gol.rs [hunk 10]
```diff
@@ -0,0 +1,314 @@
...
+    /// - Any live cell with two or three live neighbours lives on to the next generation.
+    /// - Any live cell with more than three live neighbours dies, as if by overpopulation.
+    /// - Any dead cell with exactly three live neighbours becomes a live cell, as if by reproduction.
+    fn process_cell(
+        &self,
+        next_iteration_set: &mut HashSet<(usize, usize)>,
+        current_cell: (usize, usize),
+        current_cell_alive: bool,
+    ) {
+        let (x, y) = current_cell;
```
Comment (Arkenan) [inline src/gol.rs]: With a custom struct you won't need these extractions.

File: src/gol.rs [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+                }
+            }
+        }
+        if neighbours == 3 || current_cell_alive && neighbours == 2 {
+            next_iteration_set.insert(current_cell);
+        }
+    }
+
+    /// Provides the indexes that determine a cell's neighbourhood.
+    fn get_range_for_neighbourhood(&self, cell_component: usize, limit: usize) -> (usize, usize) {
```
Comment (Arkenan) [inline src/gol.rs:106]: This is a great reusable function!

File: src/gol.rs [hunk 12]
```diff
@@ -0,0 +1,314 @@
...
+        let (y_start, y_end) = self.get_range_for_neighbourhood(current_cell.1, self.height);
+        for i in x_start..x_end {
+            for j in y_start..y_end {
+                if self.alive_cells.contains(&(i, j)) && (i, j) != (x, y) {
+                    neighbours += 1;
+                }
+            }
+        }
+        if neighbours == 3 || current_cell_alive && neighbours == 2 {
+            next_iteration_set.insert(current_cell);
```
Comment (Arkenan) [inline src/gol.rs]: Here we would just return true instead and get rid of the set in this function.

File: src/gol.rs [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+        };
+        // cell_component is supposed to work as an index either in the range 0..width or the range 0..height,
+        // in that sense start_range should be an inclusive bound and end_range should be an exclusive bound.
+        // If cell_component + 2 overflows, it means that limit = usize::MAX and cell_component = usize::MAX - 1,
+        // so end_range should be limit.
+
+        (start_range, end_range)
+    }
+
+    pub fn draw_gol_board(&self) {
```
Comment (Arkenan) [inline src/gol.rs]: This function and the following should not be part of the Game of Life struct implementation, as they are tied to the drawing library. Ideally, you should be able to give your Game of Life struct implemented should only code the game logic, and anyone should be able to implement whatever on top of it. They could implement a graphic interface, or an http api, or a console ui. Let's move these functions to a separate gol_gui.rs module.

Also, if you think about it, it's not great to have the GoL logic know about screen sizes, which is something that suggests that it should go somewhere else.

File: src/gol.rs [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+
+pub fn insert_glider_patter_middle(gol: &mut GameOfLife) {
+    if gol.width > 2 && gol.height > 2 {
+        gol.add_living_cell((gol.width / 2, gol.height / 2 + 1));
+        gol.add_living_cell((gol.width / 2 + 1, gol.height / 2 + 2));
+        gol.add_living_cell((gol.width / 2 + 1, gol.height / 2 + 3));
+        gol.add_living_cell((gol.width / 2 + 2, gol.height / 2 + 2));
+        gol.add_living_cell((gol.width / 2 + 2, gol.height / 2 + 1));
+    }
+}
```
Comment (Arkenan) [inline src/gol.rs]: These three functions are a bit too specific. If we want to encode some common patterns, which is fine, we should have functions like:

- `insert_line(gol, position, length)`
- `insert_square(gol, position, size)`
- `insert_glider(gol, position)`

This would allow you to have a more configurable initial game (either with constants in main, or with files) and also more tests with edge cases or different positions.

File: src/gol.rs [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+        // in that sense start_range should be an inclusive bound and end_range should be an exclusive bound.
+        // If cell_component + 2 overflows, it means that limit = usize::MAX and cell_component = usize::MAX - 1,
+        // so end_range should be limit.
+
+        (start_range, end_range)
+    }
+
+    pub fn draw_gol_board(&self) {
+        let x_padding = screen_width() / self.width as f32;
+        let y_padding = screen_height() / self.height as f32;
```
Comment (Arkenan) [inline src/gol.rs]: This is not padding, this is the cell width and height. Padding is the space between two objects. Consider having a GolUI struct like so:

```rust
struct GolUI {
  x_size,
  y_size
}
```

And then have it implement the drawing functions. This way you don't have to "calculate" those widths everywhere.

File: src/gol.rs [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+        for i in 0..self.width {
+            for j in 0..self.height {
+                if self.alive_cells.contains(&(i, j)) {
+                    draw_rectangle(
+                        (i as f32 * x_padding) + 1.0,
+                        (j as f32 * y_padding) + 1.0,
+                        x_padding - 1.0,
+                        y_padding - 1.0,
+                        BLACK,
+                    );
```
Comment (Arkenan) [inline src/gol.rs]: After implementing the GolUI you can benefit from having a `gol_ui.draw_cell(cell, is_alive)` method.

File: src/gol.rs [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+                }
+            }
+        }
+    }
+
+    pub fn draw_or_drop_new_cell(&mut self, position: (f32, f32)) {
+        let x_padding = screen_width() / self.width as f32;
+        let y_padding = screen_height() / self.height as f32;
+        let x_position = (position.0 / x_padding) as usize;
+        let y_position = (position.1 / y_padding) as usize;
```
Comment (Arkenan) [inline src/gol.rs]: Consider implementing `ui.get_cell(mouse_position: (f32, f32)) -> Cell`

File: src/gol.rs [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+            draw_rectangle(
+                (x_position as f32 * x_padding) + 1.0,
+                (y_position as f32 * y_padding) + 1.0,
+                x_padding - 1.0,
+                y_padding - 1.0,
+                BLACK,
+            );
+        }
+    }
+}
```
Comment (Arkenan) [inline src/gol.rs]: No need to draw it here. It will be drawn on the next iteration. But also, let's have a 

```rust
gol.toggle(cell)
```

function to change the state of a cell.

File: src/gol.rs [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+        let (y_start, y_end) = gol.get_range_for_neighbourhood(current_cell.1, gol.height);
+        assert_eq!(x_start, 3);
+        assert_eq!(x_end, 5);
+        assert_eq!(y_start, 3);
+        assert_eq!(y_end, 5);
+    }
+
+    #[test]
+    fn range_of_neighbourhood_overflow_case() {
+        let mut gol = GameOfLife::new(usize::MAX, usize::MAX);
```
Comment (Arkenan) [inline src/gol.rs:228]: Impressive that you got this to work with a max size grid!

File: src/gol.rs [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+            (_, true) => cell_component,
+        };
+        let end_range = match (cell_component).overflowing_add(2) {
+            (n, false) => min(n, limit),
+            (_, true) => limit,
+        };
+        // cell_component is supposed to work as an index either in the range 0..width or the range 0..height,
+        // in that sense start_range should be an inclusive bound and end_range should be an exclusive bound.
+        // If cell_component + 2 overflows, it means that limit = usize::MAX and cell_component = usize::MAX - 1,
+        // so end_range should be limit.
```
Comment (Arkenan) [inline src/gol.rs:114]: This is a great comment, it's a very confusing thing to read the 2 instead of the 1 without clarification. I would put it before the `let end_range = ...` line though, as it's usually better to read the comment before you read the thing that's being commented on.

File: src/gol.rs (hunk 7 above)
Comment (cdiielsi) [reply src/gol.rs]: I decided to use HashSet in case that in a new iteration I had to make the board's size variable and because I thought that given the game's rule of "dying because of overpopulation" the amount of living cells tends to be small compared to the board's size, so my idea was to save space by only keeping track of the living cells with the set.
I'm documenting this on the README and I'm implementing the Cell type now.

File: src/gol.rs (hunk 8 above)
Comment (cdiielsi) [reply src/gol.rs]: Roger that! Fixing it right away!

File: src/gol.rs (hunk 14 above)
Comment (cdiielsi) [reply src/gol.rs]: I'll fix it! The original idea behind this functions was to just use them as an easy set up for testing, that's why I made them so specific.

File: src/gol.rs [hunk 21]
```diff
@@ -77,144 +108,136 @@ impl GameOfLife {
...
+    })?;
+    gol.add_living_cell(Cell {
+        x: position.x + 2,
+        y: position.y,
+    })?;
+    gol.add_living_cell(Cell {
+        x: position.x + 2,
+        y: position.y + 1,
+    })?;
+    Ok(())
```
Comment (Arkenan) [inline src/gol.rs]: Code is good and perfectly readable as it is now, but I want to propose a couple of refactors that might be interesting to think about as an exercise.

This code could be simplified if we had a `cell.add(x,y) -> Cell`, method. It would read something like

```rust
    gol.add_living_cell(position.add(1,1))?;
    gol.add_living_cell(position.add(1,2))?;
    gol.add_living_cell(position.add(2,0)?;
    gol.add_living_cell(position.add(2,1))?;
```

In general, if something repeats itself three times or more, it's probably worth thinking about generalizing it.

Aside from adding, we can see that there's a different thing that repeats: drawing different positions. We could do something like this:

```rust
let positions = [Cell::new(1,1), Cell::new(1,2), Cell::new(2,0), Cell::new(2,1)]
for position in positions {
  gol.add_living_cell(base.add(position))
}
```

This could even be generalized further. What's common in every figure? What makes them differently? They all have a bunch of positions and they just have to be drawn offseted by some base.

```rust
fn draw_figure(gol, base, positions) {
  for position in positions {
    gol.add_living_cell(base.add(position))
  }
}

fn draw_glider_pattern(gol, base) {
  draw_figure(gol, base, [Cell::new(1,1), Cell::new(1,2), Cell::new(2,0), Cell::new(2,1)]);
}
```

and the same for others. It's worth thinking at each step "Am I going too far?" and the things to take into account are always:
- Am I deleting actual repetition?
- What is the same and what is different between each repetition?
- Am I preparing myself for non-existent problems by over-generalizing?

File: src/gol_gui.rs [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+        } else {
+            draw_rectangle(
+                (cell.x as f32 * self.x_size) + 1.0,
+                (cell.y as f32 * self.y_size) + 1.0,
+                self.x_size - 1.0,
+                self.y_size - 1.0,
+                WHITE,
+            );
+        }
+    }
```
Comment (Arkenan) [inline src/gol_gui.rs]: Small correction here: instead of doing this twice we could just:

```rust
let color = if is_alive {BLACK} else {WHITE};
draw_rectangle(
  (cell.x as f32 * self.x_size) + 1.0,
  (cell.y as f32 * self.y_size) + 1.0,
  self.x_size - 1.0,
  self.y_size - 1.0,
  color,
);
```

What is different? The color. What is the same? Everything else. Take what's different out. Do the rest once.

**Review by Arkenan — APPROVED** [review]
Code looks great, LGTM

File: src/gol.rs (hunk 21 above)
Comment (cdiielsi) [reply src/gol.rs]: I see!! Thank you so much!! I guess I didn't consider generalizing this code because I was only using it for testing at first, and since I chose small patterns I even decided to have them hardcoded at the very beginning. But I guess once I used this functions to initialize the interactive game I should have made a refactor like this. An idea I had for a possible future iteration was to have the game start with a concrete picture so you watch the game develop from that starting point, to do that I think this refactor would be really necessary.
